### PR TITLE
VR Dynamic Month View - step 1

### PR DESCRIPTION
**Pull request by lucatume**
Tickets:

* https://central.tri.be/issues/129407
* https://central.tri.be/issues/129405

This PR adds some base utils required for the Month view dynamic implementation:
* the `Tribe\Utils\Collection_Interface`; anything that behaves like a collection should implement it.
* the `Tribe\Utils\Collection_Trait` to allow the implementation of the `Tribe\Utils\Collection_Interface` implementing just one method (the `all()`) one.
* the `Tribe\Utils\Lazy_Collection`; does what it says by providing something that behaves like a collection (and an array) that will fetch its items if really required to.
* the `Tribe\Utils\Post_Thumbnail; it's a class wrapping a post thumbnail information (URLs, srcset and the like) in a lazy-fetching wrapper.

### CODE REVIEW

File: src/Tribe/Utils/Post_Thumbnail.php [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+			return [];
+		}
+
+		$thumbnail_data = array_combine(
+			$image_sizes,
+			array_map( static function ( $size ) use ( $thumbnail_id ) {
+				$size_data = wp_get_attachment_image_src( $thumbnail_id, $size );
+
+				if(false === $size_data){
+					return (object)[
```
Comment (tr1b0t) [inline src/Tribe/Utils/Post_Thumbnail.php]: * :no_entry_sign: There must be a space between if and the first parenthesis '('
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space after opening parenthesis is prohibited
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space before closing parenthesis is prohibited
* :no_entry_sign: You must have a space after a closing parenthesis

<sub>hash: a02b1439e9c166c77dffd55a231d0eab</sub>

File: src/Tribe/Utils/Post_Thumbnail.php [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+						'url' => '',
```
Comment (tr1b0t) [inline src/Tribe/Utils/Post_Thumbnail.php]: :no_entry_sign: Expected 1 space(s) after cast statement; 0 found

<sub>hash: 5cce090f1ba47615ae0591732a320405</sub>

File: src/Tribe/Utils/Post_Thumbnail.php [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+						'is_intermediate' => false,
+					];
+				}
+
+				return (object) [
+					'url'             => Arr::get( $size_data, 0, '' ),
+					'width'           => Arr::get( $size_data, 1, '' ),
+					'heigth'          => Arr::get( $size_data, 2, '' ),
+					'is_intermediate' => (bool)Arr::get( $size_data, 3, false ),
+				];
```
Comment (tr1b0t) [inline src/Tribe/Utils/Post_Thumbnail.php]: :no_entry_sign: Expected 1 space(s) after cast statement; 0 found

<sub>hash: 714a39768b31e2d8c2383a9cf7813272</sub>

**Comment by lucatume** [comment]
Holding this while waiting for 19.10 to come live.

**Comment by juanfra** [comment]
Looking great!

**Pull request merged**